### PR TITLE
[14.0][IMP] mrp_subcontracting_skip_no_negative: Show negative stock error of subcontracting component

### PR DESCRIPTION
**Pull request by victoralmau**
Show negative stock error of subcontracting component

@Tecnativa TT50668

### CODE REVIEW

**Comment by victoralmau** [comment]
@pedrobaeza What do you think of this approach? I clarify that I don't like it, but I can't find another way to have the error (if applicable) of negative quantity displayed before done of the subcontracting receipt.

**Review by pedrobaeza — COMMENTED** [review]
Yeah, we are replacing business logic here. What you have to do is to manually check the components lines and issue the exception with the proper message if one of them doesn't comply with the needed quantity. Only adding checks, not replacing behavior.

**Comment by victoralmau** [comment]
> Yeah, we are replacing business logic here. What you have to do is to manually check the components lines and issue the exception with the proper message if one of them doesn't comply with the needed quantity. Only adding checks, not replacing behavior.

Changed the approach.

**Comment by victoralmau** [comment]
Ready to review @pedrobaeza

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
@AungKoKoLin1997 can you check?

File: mrp_subcontracting_skip_no_negative/models/stock_move.py [hunk 1]
```diff
@@ -35,6 +38,43 @@ def _action_done(self, cancel_backorder=False):
...
+            for p_move in unassigned_productions.move_raw_ids.filtered(
+                lambda x: x.state != "assigned"
+                and not x.product_id.allow_negative_stock
+                and not x.product_id.categ_id.allow_negative_stock
+                and not x.location_id.allow_negative_stock
+            ):
+                product = p_move.product_id.sudo()
+                location = p_move.location_id
+                location_qty = (
+                    product.sudo().with_context(location=location.id).free_qty
```
Comment (AungKoKoLin1997) [inline mrp_subcontracting_skip_no_negative/models/stock_move.py]: The use of sudo() is redundant. Remove it.

File: mrp_subcontracting_skip_no_negative/models/stock_move.py [hunk 2]
```diff
@@ -35,6 +38,41 @@ def _action_done(self, cancel_backorder=False):
...
+                location = p_move.location_id
+                location_qty = product.with_context(location=location.id).free_qty
+                new_qty = location_qty - p_move.product_uom_qty
+                if float_compare(new_qty, 0, precision_digits=qty_precision) == -1:
+                    raise ValidationError(
+                        _(
+                            "You cannot validate this stock operation because the "
+                            "stock level of the product '%s' would become negative "
+                            "(%s) on the stock location '%s' and negative stock is "
+                            "not allowed for this product and/or location."
```
Comment (AungKoKoLin1997) [inline mrp_subcontracting_skip_no_negative/models/stock_move.py]: I think `stock level of the component product` is clearer than just `stock level of the product`, as this product isn’t in the move line. Specifying it as a component product helps avoid confusion. How do you think?

**Review by AungKoKoLin1997 — APPROVED** [review]
LGTM.

**Comment by AungKoKoLin1997** [comment]
Can you please forward port to another versions?

**Pull request merged**